### ISSUE TITLE
Lag between logging sources and catalog changes

### ISSUE THREAD

**Issue by ruchirK** [MaterializeInc/materialize#2576]
We recently introduced `mz_kafka_sinks` and `mz_avro_ocf_sinks` as two logging sources that maintain metadata for kafka and avro ocf sinks respectively. However, these (and potentially other logging sources) have the potential to be inconsistent with catalog changes -- specifically if a catalog entry is deleted and reinserted, we can repeatedly see stale metadata that should have been deleted.

I wasn't able to reproduce this manually (perhaps I'm too slow of a typer) but I can reliably reproduce this with this diff to the avro ocf testdrive test
```
altaria-2:materialize Test$ git diff
diff --git a/test/testdrive/avro-ocf.td b/test/testdrive/avro-ocf.td
index 573de754..d94b03f3 100644
--- a/test/testdrive/avro-ocf.td
+++ b/test/testdrive/avro-ocf.td
@@ -165,19 +165,19 @@ ts            NO      timestamp
 ts_tz         NO      timestamp
 mz_obj_no     NO      int8

-> CREATE SINK basic_sink_${testdrive.seed} FROM basic
+> CREATE SINK basic_sink FROM basic
   INTO AVRO OCF '${testdrive.temp-dir}/basic-sink.ocf'

-$ avro-ocf-verify sink=materialize.public.basic_sink_${testdrive.seed}
+$ avro-ocf-verify sink=materialize.public.basic_sink
 {"before": null, "after": {"a": 1, "b": 2, "mz_obj_no": 1}}
 {"before": null, "after": {"a": 3, "b": 4, "mz_obj_no": 2}}

 > CREATE VIEW dateish AS
   SELECT d FROM timestamp_source

-> CREATE SINK date_sink_${testdrive.seed} FROM dateish
+> CREATE SINK date_sink FROM dateish
   INTO AVRO OCF '${testdrive.temp-dir}/date-sink.ocf'

-$ avro-ocf-verify sink=materialize.public.date_sink_${testdrive.seed}
+$ avro-ocf-verify sink=materialize.public.date_sink
 {"before": null, "after": {"d": 10988}}
 {"before": null, "after": {"d": 10957}}
altaria-2:materialize Test$
```

This diff changes it so sink names are identical across repeat runs of the test (as they should be)

When you run this test it will pass the first time but fail the second time with an error like

```
test/testdrive/avro-ocf.td:171:1: error: reading sink file /var/folders/jp/0_nyxpyj4vxg8x4c1kx1vhl80000gn/T/.tmplwBXdj/basic-sink-u46-1586210039-8060089200439147410.ocf: No such file or directory (os error 2)
     |
 170 |
 171 | $ avro-ocf-verify sink=materialize.public.basic_sink
     | ^
altaria-2:materialize Test$ git diff
```

Note the `basic-sink-u46` in the file name -- this is telling us that its mapping a sink named `basic-sink` to global id `u46`

but if you look at the log output you will see that `basic-sink` was dropped along with its source

```
Apr 06 17:55:20.148  INFO catalog: drop index materialize.public.basic_primary_idx (u37)
Apr 06 17:55:20.148  INFO catalog: drop sink materialize.public.basic_sink (u46)
Apr 06 17:55:20.148  INFO catalog: drop source materialize.public.basic (u36)
```

and currently `basic-sink` is bound to a different global id

```
Apr 06 17:55:20.839  INFO catalog: create sink materialize.public.basic_sink (u59)
```

**Comment by frankmcsherry** [comment]
Related to #913.

Logging sources weren't designed to be consistent with the catalog (they predate it).  There might be a bug, but it could plausibly be in the expectations around logging sources. Issue #913 was about mirroring catalog information in the logging channels, though we probably need to be careful about which logging sources reflect potentially consistent information and which ones just reflect worker state at worker-local timestamps.

**Comment by benesch** [comment]
Yeah, sorry, this is more of a UX "bug" than an actual code bug. It is both expected and unfortunate!

**Comment by brandonbloom** [comment]
I've seen _some_ of the discussion around this and am confident you're all thinking about the rights to improve this. Just wanted to add a note about how I intend to use the catalog and why I came looking at these issues:

I'm doing some automation for testing, etc. and was writing a little tool to automatically tail some output logs for a test run. I realized that querying for the catalog and avro sinks file paths is very, very slow, shortly after issuing some creates. In fact, sometimes it seems to block indefinitely (several minutes before I give up), which may indicate some sort of synchronization bug as well.

So big +1 for these things behavior as nice and fast and as consistently as all the normal materialized views. Thanks!

**Comment by benesch** [comment]
Hmm, @brandonbloom, that sounds like something else is going very wrong. It shouldn't take more than a second for the catalog view to reflect the new sink. I'd expect tens of milliseconds, definitely not minutes.

I think it's quite likely that one of your other CREATEs caused pathological behavior (perhaps a bad query plan) that stalled out the system. Next time that happens, is there any chance you'd be able to send along the views that caused the issue? If there's nothing sensitive in your view/source/sink definitions, you can just send along the `mzdata/catalog` file if you like.

**Comment by brandonbloom** [comment]
@benesch I've emailed bugs@ with more details. Thanks!

**Comment by benesch** [comment]
As an example of the kind of pain this causes, there's presently a lot of code in the billing demo to handle the inconsistent state of the catalog: https://github.com/MaterializeInc/materialize/blob/72b394d4f30c1f44dc0fe4475814bd777364d9e1/demo/billing/src/mz_client.rs#L86-L217

That code isn't even defensive enough; there's one more missing retry loop that I'm adding right now to fix CI flakiness.

The importance of this issue and #913 is becoming increasingly clear to me. I think it's an important proactive step we need to take ASAP; it'll also unblock a lot of ORM compatibility work that's on the horizon that will require `pg_catalog` tables.

**Comment by cuongdo** [comment]
@justinj has started looking at this. The impact on ORM / GraphQL compatibility is especially noteworthy, making this a high-priority.

**Comment by benesch** [comment]
I think we should consider this issue closed when the following sequence of queries works:

```sql
CREATE SINK snk FROM src INTO KAFKA ...
SELECT topic FROM mz_kafka_sinks NATURAL JOIN mz_catalog_names WHERE name = 'materialize.public.snk'
```

If you run that today, and run the queries back to back quickly enough, the second query won't return any results, and you'll have to run it again a few milliseconds later. (If you're _really_ unlucky, and run `DROP SINK snk; CREATE SINK ...; SELECT topic ...`, you'll seem to get a topic, but it will be the topic for the _last_ iteration of the sink, not the new iteration.)

As mentioned elsewhere I don't think the problem here is that the logging views are laggy but that `mz_kafka_sinks` and `mz_catalog_names` are logging views at all. Logging views should continue to operate under whatever consistency model makes sense for them—we should just get the stuff that requires more consistency out of the logging framework.

**Comment by justinj** [comment]
Fixed (for `mz_kafka_sinks`) in #3535.

**Comment by benesch** [comment]
Sorry to do this 🙈 but I'm going to reopen briefly to track fixing `mz_catalog_names` (which I'm working on now).
